presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -13,75 +13,82 @@
 
 import static io.prestosql.plugin.prometheus.PrometheusErrorCode.PROMETHEUS_TABLES_METRICS_RETRIEVE_ERROR;
 import static io.prestosql.plugin.prometheus.PrometheusErrorCode.PROMETHEUS_UNKNOWN_ERROR;
 import static io.prestosql.spi.StandardErrorCode.NOT_FOUND;
 import static io.prestosql.spi.type.TypeSignature.mapType;
 import static io.prestosql.spi.type.VarcharType.VARCHAR;
 import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.Files.readString;
 import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 public class PrometheusClient
 {
-    private PrometheusConnectorConfig config;
+    static final Type TIMESTAMP_COLUMN_TYPE = TimestampType.createTimestampType(3);
     static final String METRICS_ENDPOINT = "/api/v1/label/__name__/values";
-    private static final OkHttpClient httpClient = new OkHttpClient.Builder().build();
+    private static final OkHttpClient httpClient = new Builder().build();
+
+    private final Optional<File> bearerTokenFile;
     private final Supplier<Map<String, Object>> tableSupplier;
-    private final TypeManager typeManager;
+    private final Type varcharMapType;
 
     @Inject
     public PrometheusClient(PrometheusConnectorConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
-            throws URISyntaxException
     {
         requireNonNull(config, "config is null");
         requireNonNull(metricCodec, "metricCodec is null");
         requireNonNull(typeManager, "typeManager is null");
 
-        tableSupplier = Suppliers.memoizeWithExpiration(metricsSupplier(metricCodec, getPrometheusMetricsURI(config)),
-                (long) config.getCacheDuration().getValue(), config.getCacheDuration().getUnit());
-        this.typeManager = typeManager;
-        this.config = config;
+        bearerTokenFile = config.getBearerTokenFile();
+        URI prometheusMetricsUri = getPrometheusMetricsURI(config.getPrometheusURI());
+        tableSupplier = Suppliers.memoizeWithExpiration(
+                () -> fetchMetrics(metricCodec, prometheusMetricsUri),
+                config.getCacheDuration().toMillis(),
+                MILLISECONDS);
+        varcharMapType = typeManager.getType(mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()));
     }
 
-    private URI getPrometheusMetricsURI(PrometheusConnectorConfig config)
-            throws URISyntaxException
+    private static URI getPrometheusMetricsURI(URI prometheusURI)
     {
-        // endpoint to retrieve metric names from Prometheus
-        URI uri = config.getPrometheusURI();
-        return new URIBuilder()
-                .setScheme(uri.getScheme())
-                .setHost(uri.getAuthority())
-                .setPath(uri.getPath().concat(METRICS_ENDPOINT))
-                .build();
+        try {
+            // endpoint to retrieve metric names from Prometheus
+            return new URIBuilder()
```

Comment:
We should get rid of this usage of `org.apache.http.client.utils.URIBuilder`. There's another PR #4698 to use Airlift HTTP client.
```java
return new URI(uri.getScheme(), uri.getAuthority(), uri.getPath() + METRICS_ENDPOINT, null, null);